src/main/kotlin/brokers/PgBroker.kt

```diff
@@ -0,0 +1,153 @@
+    val attemptNum = integer("attempt_num").default(0)
+    val scheduledAt = timestamp("scheduled_at").defaultExpression(CurrentTimestamp())
+    val startedAt = timestamp("started_at").nullable().default(null)
+
+    override val primaryKey = PrimaryKey(id)
+
+    init {
+        index("message_selection", false, queueName, scheduledAt, startedAt)
+    }
+}
+
+private fun Message.getId() = UUID.fromString(this.headers[PgBroker.ID_HEADER])
+private fun Message.getCallId() = UUID.fromString(this.headers["call-id"])
+private fun Message.getAttemptNum() = Integer.parseInt(this.headers.getOrDefault("attempt-num", "0"))
+
+class PgBroker(
+    val dbWrapper: DbWrapper,
+    val scope: CoroutineScope = GlobalScope,
+
+    val emptyMessageDelay: Duration = 5000.milliseconds,
+    val messageReservationTimeoutMs: Duration = 5.minutes,
+): IMessageBroker {
+    companion object {
+        const val HEADERS_PREFIX = "kot-pg"
+        const val ID_HEADER = "${HEADERS_PREFIX}-id"
+    }
+
+    init {
+        transaction(dbWrapper.connection) {
+            SchemaUtils.createMissingTablesAndColumns(KotaskMessages)
+        }
+    }
+
+    override fun submitMessage(queueName: QueueName, message: Message): Unit {
+
+        transaction(dbWrapper.connection) {
+            val id = UUID.randomUUID()
+            KotaskMessages.insert {
+                it[this.id] = id
+                it[this.callId] = message.getCallId()
+                it[this.queueName] = queueName
+                it[this.body] = message.body
+                it[this.attemptNum] = message.getAttemptNum()
+                it[this.delayMs] = message.delayMs
+                it[this.scheduledAt] = Instant.now().plusMillis(message.delayMs)
+                it[this.headers] = Json.encodeToString(
+                    buildMap {
+                        putAll(message.headers)
+                        put(ID_HEADER, id.toString())
+                    }
+                ).encodeToByteArray()
+            }
+        }
+    }
+
+    override fun startConsumer(queueName: QueueName, handler: ConsumerHandler): IConsumer {
+        var isStopped = false
+        val job = scope.launch(MDCContext()) {
+            while (!isStopped) {
+                val messages = transaction(dbWrapper.connection) {
+                    val messages = KotaskMessages
+                        .selectAll()
+                        .andWhere {
+                            (KotaskMessages.queueName eq queueName) and
+                            (KotaskMessages.scheduledAt less Instant.now()) and
+                            (
+                                KotaskMessages.startedAt less Instant.now().minusSeconds(messageReservationTimeoutMs.inWholeSeconds) or // settings
+                                KotaskMessages.startedAt.isNull()
+                            )
+                        }
+                        .limit(100) // settings
+                        .forUpdate()
```

Comment:
I'm not fully understand `forUpdate`. Will it lock the record for being selected by other processes? I'm thinking about how multiple simultaneous consumers will behave: are they going to select the same messages or not?